File: cache-warmer/cache-warmer.py
import asyncio
import signal
import httpx
import time
from datetime import datetime, timezone
from defusedxml import ElementTree as ET

# CONFIGURATION
SITEMAP_URL = "https://pbservices.ge/sitemap-index.xml"
CONCURRENT_REQUESTS = 1
REQUEST_DELAY_SECONDS = 2.0
RUN_AT_HOUR_UTC = 3  # Fixed UTC hour for daily run (0-23)
CUSTOM_USER_AGENT = "SevallaCacheWarmerSafe-SecureToken-99x"
DISCOVERY_CONCURRENCY = 3  # Max simultaneous sitemap requests
RETRY_COUNT = 2  # Retries per URL on transient failure

# Global flags
_shutdown = False
_force_run = False

def log(msg):
    """Timestamped print."""
    ts = datetime.now(timezone.utc).isoformat()
    print(f"[{ts}] {msg}")


def handle_shutdown(sig, frame):
    """Set shutdown flag on SIGTERM/SIGINT."""
    global _shutdown
    log(f"Received signal {sig}, shutting down gracefully...")
    _shutdown = True


def handle_force_run(sig, frame):
    """Set force-run flag on SIGUSR1."""
    global _force_run
    log("Received SIGUSR1, forcing immediate cache warm cycle.")
    _force_run = True


signal.signal(signal.SIGTERM, handle_shutdown)
signal.signal(signal.SIGINT, handle_shutdown)
signal.signal(signal.SIGUSR1, handle_force_run)

discovery_semaphore = asyncio.Semaphore(DISCOVERY_CONCURRENCY)


async def fetch_and_parse_sitemap(client, target_url, urls_set):
    """Recursively parses sitemap indexes and extracts flat page URLs."""
    async with discovery_semaphore:
        try:
            response = await client.get(target_url, timeout=15.0)
            response.raise_for_status()

            root = ET.fromstring(response.content)
            root_tag = root.tag.split("}")[-1] if "}" in root.tag else root.tag

            locs = []
            for elem in root.iter():
                tag = elem.tag.split("}")[-1] if "}" in elem.tag else elem.tag
                if tag == "loc" and elem.text:
                    locs.append(elem.text.strip())

            if root_tag == "sitemapindex":
                log(f"[Index] Scanning directory: {target_url}")
                tasks = [fetch_and_parse_sitemap(client, loc, urls_set) for loc in locs]
                await asyncio.gather(*tasks)

            elif root_tag == "urlset":
                for loc in locs:
                    urls_set.add(loc)

        except Exception as e:
            log(f"[ERROR] Failed processing sitemap target {target_url}: {e}")


async def warm_url(client, url):
    """Hits a single URL with retries. Reads cache state from headers."""
    for attempt in range(1 + RETRY_COUNT):
        if _shutdown:
            return
        try:
            start_time = time.time()
            response = await client.get(url, timeout=10.0)
            duration = time.time() - start_time

            kinsta_cache = response.headers.get("X-Kinsta-Cache", "UNKNOWN")
            edge_cache = response.headers.get("Ki-Cf-Cache-Status", "UNKNOWN")

            log(f"[{response.status_code}] Kinsta: {kinsta_cache} | Edge: {edge_cache} | Time: {duration:.2f}s -> {url}")
            return  # success, exit retry loop
        except Exception as e:
            if attempt < RETRY_COUNT:
                backoff = 2 ** attempt
                log(f"[RETRY {attempt+1}/{RETRY_COUNT}] {url} — {e}. Waiting {backoff}s...")
                await asyncio.sleep(backoff)
            else:
                log(f"[ERROR] Failed to warm page {url}: {e}")


async def run_warmer():
    headers = {"User-Agent": CUSTOM_USER_AGENT}

    async with httpx.AsyncClient(headers=headers, follow_redirects=True) as client:
        log("--- Starting Sitemap Discovery Phase ---")
        urls_set = set()
        await fetch_and_parse_sitemap(client, SITEMAP_URL, urls_set)

        all_pages = list(urls_set)
        total_pages = len(all_pages)

        if not total_pages:
            log("[Warning] No page URLs discovered.")
            return

        log(f"--- Discovery Finished: Unique Pages Found: {total_pages} ---")
        log("--- Beginning Safe Sequential Warming Loop ---")

        for url in all_pages:
            if _shutdown:
                log("Shutdown requested, stopping warming early.")
                break
            await warm_url(client, url)
            await asyncio.sleep(REQUEST_DELAY_SECONDS)

        log("Cache warming cycle completed cleanly.")


def _seconds_until_next_run(hour_utc):
    """Seconds until the next occurrence of hour_utc (0-23) in UTC."""
    now = datetime.now(timezone.utc)
    target = now.replace(hour=hour_utc, minute=0, second=0, microsecond=0)
    if target <= now:
        target = target.replace(day=now.day + 1)  # handles month rollover
    return (target - now).total_seconds()


async def main():
    # Run immediately on startup
    try:
        await run_warmer()
    except Exception as e:
        log(f"[FATAL] Warmer cycle crashed: {e}")

    while True:
        wait_seconds = _seconds_until_next_run(RUN_AT_HOUR_UTC)
        log(f"Next run scheduled at {RUN_AT_HOUR_UTC:02d}:00 UTC "
            f"({wait_seconds / 3600:.1f} hours from now). "
            f"Send SIGUSR1 (kill -USR1 {__import__('os').getpid()}) to force run.")

        # Sleep in 30-second chunks so SIGUSR1 is responsive
        waited = 0.0
        while waited < wait_seconds and not _shutdown and not _force_run:
            chunk = min(30, wait_seconds - waited)
            await asyncio.sleep(chunk)
            waited += chunk

        if _shutdown:
            break

        if _force_run:
            global _force_run
            _force_run = False
            log("Manual trigger — starting cache warm cycle now.")

        try:
            await run_warmer()
        except Exception as e:
            log(f"[FATAL] Warmer cycle crashed: {e}")


if __name__ == "__main__":
    asyncio.run(main())

File: services/nodejs/src/cache-warmer.js
const { XMLParser } = require("fast-xml-parser");
const fs = require("fs");
const path = require("path");

// CONFIGURATION
const SITEMAP_URL = "https://pbservices.ge/sitemap-index.xml";
const REQUEST_DELAY_MS = 2000; // 2s between requests
const DISCOVERY_CONCURRENCY = 3;
const RETRY_COUNT = 2;
const USER_AGENT = "SevallaCacheWarmerSafe-SecureToken-99x";
const SUMMARY_FILE = path.join(__dirname, "..", "cache-warmer-last-run.json");

const parser = new XMLParser({ ignoreAttributes: false });

// ── Concurrency guard: prevent overlapping runs ──
let isRunning = false;

// Concurrency limiter for sitemap discovery
class Semaphore {
  constructor(max) {
    this.max = max;
    this.running = 0;
    this.queue = [];
  }
  async acquire() {
    if (this.running < this.max) {
      this.running++;
      return;
    }
    return new Promise((resolve) => this.queue.push(resolve));
  }
  release() {
    this.running--;
    const next = this.queue.shift();
    if (next) {
      this.running++;
      next();
    }
  }
}

const discoverySem = new Semaphore(DISCOVERY_CONCURRENCY);

function log(msg) {
  console.log(`[${new Date().toISOString()}] ${msg}`);
}

function extractLocs(root) {
  const locs = [];
  function walk(node) {
    if (!node || typeof node !== "object") return;
    const keys = Object.keys(node);
    for (const key of keys) {
      if (key.endsWith("loc") && typeof node[key] === "string") {
        locs.push(node[key].trim());
      } else if (typeof node[key] === "object") {
        walk(node[key]);
      } else if (Array.isArray(node[key])) {
        node[key].forEach((item) => walk(item));
      }
    }
  }
  walk(root);
  return locs;
}

async function fetchSitemap(url, urlsSet) {
  await discoverySem.acquire();
  try {
    const res = await fetch(url, {
      headers: { "User-Agent": USER_AGENT },
      signal: AbortSignal.timeout(15000),
    });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    const text = await res.text();
    const parsed = parser.parse(text);

    const rootKey = Object.keys(parsed).find(
      (k) => k.includes("sitemapindex") || k.includes("urlset")
    );

    if (rootKey && rootKey.includes("sitemapindex")) {
      log(`[Index] Scanning directory: ${url}`);
      const locs = extractLocs(parsed[rootKey]);
      const results = await Promise.allSettled(
        locs.map((loc) => fetchSitemap(loc, urlsSet))
      );
      results.forEach((r, i) => {
        if (r.status === "rejected")
          log(`[ERROR] Sub-sitemap failed: ${locs[i]} — ${r.reason}`);
      });
    } else if (rootKey && rootKey.includes("urlset")) {
      const locs = extractLocs(parsed[rootKey]);
      locs.forEach((loc) => urlsSet.add(loc));
    }
  } catch (e) {
    log(`[ERROR] Failed processing sitemap ${url}: ${e.message}`);
  } finally {
    discoverySem.release();
  }
}

/**
 * Warms a single URL. Returns { ok, status, error, kinsta, cdn, edge }.
 */
async function warmUrl(url) {
  for (let attempt = 0; attempt <= RETRY_COUNT; attempt++) {
    try {
      const start = Date.now();
      const res = await fetch(url, {
        headers: { "User-Agent": USER_AGENT },
        signal: AbortSignal.timeout(10000),
      });
      const duration = ((Date.now() - start) / 1000).toFixed(2);
      const kinsta = res.headers.get("X-Kinsta-Cache") || "UNKNOWN";
      const cdn = res.headers.get("CF-Cache-Status") || "UNKNOWN";
      const edge = res.headers.get("Ki-Cf-Cache-Status") || "UNKNOWN";
      log(`[${res.status}] Kinsta: ${kinsta} | CDN: ${cdn} | Edge: ${edge} | Time: ${duration}s -> ${url}`);
      return { ok: true, status: res.status, kinsta, cdn, edge };
    } catch (e) {
      if (attempt < RETRY_COUNT) {
        const backoff = 2 ** attempt;
        log(`[RETRY ${attempt + 1}/${RETRY_COUNT}] ${url} — ${e.message}. Waiting ${backoff}s...`);
        await sleep(backoff * 1000);
      } else {
        log(`[ERROR] Failed to warm page ${url}: ${e.message}`);
        return { ok: false, status: null, error: e.message, url };
      }
    }
  }
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ── Stats helpers ──

function initStats() {
  return { hit: 0, miss: 0, bypass: 0, unknown: 0 };
}

function tallyStats(stats, value) {
  const v = (value || "").toLowerCase();
  if (v === "hit") stats.hit++;
  else if (v === "miss") stats.miss++;
  else if (v === "bypass") stats.bypass++;
  else stats.unknown++;
}

function formatStats(label, stats, total) {
  const pct = (n) => total > 0 ? ` (${((n / total) * 100).toFixed(1)}%)` : "";
  return `${label}: ${stats.hit} HIT${pct(stats.hit)}, ${stats.miss} MISS${pct(stats.miss)}, ${stats.bypass} BYPASS${pct(stats.bypass)}` +
    (stats.unknown ? `, ${stats.unknown} UNKNOWN` : "");
}

async function runWarmer() {
  if (isRunning) {
    log("[SKIP] Warmer is already running — concurrent run prevented.");
    return;
  }
  isRunning = true;

  const startTime = new Date().toISOString();
  // Stats accumulators
  const kinstaStats = initStats();
  const cdnStats = initStats();
  const edgeStats = initStats();
  const failedUrls = [];

  try {
    log("--- Starting Sitemap Discovery Phase ---");
    const urlsSet = new Set();
    await fetchSitemap(SITEMAP_URL, urlsSet);

    const allPages = [...urlsSet];
    if (!allPages.length) {
      log("[Warning] No page URLs discovered.");
      return;
    }

    log(`--- Discovery Finished: Unique Pages Found: ${allPages.length} ---`);
    log("--- Beginning Safe Sequential Warming Loop ---");

    for (const url of allPages) {
      const result = await warmUrl(url);
      if (result.ok) {
        tallyStats(kinstaStats, result.kinsta);
        tallyStats(cdnStats, result.cdn);
        tallyStats(edgeStats, result.edge);
      } else {
        failedUrls.push({ url, error: result.error });
      }
      await sleep(REQUEST_DELAY_MS);
    }

    // ── Summary ──
    const endTime = new Date().toISOString();
    const total = allPages.length;
    const ok = total - failedUrls.length;
    const fail = failedUrls.length;

    const summaryLines = [
      "",
      "═══════════════════════════════════════════",
      "           CACHE WARMER — SUMMARY           ",
      "═══════════════════════════════════════════",
      `Started:     ${startTime}`,
      `Finished:    ${endTime}`,
      `Total URLs:  ${total}`,
      `Successful:  ${ok}`,
      `Failed:      ${fail}`,
      "",
      "── Cache Status ──",
      formatStats("Kinsta", kinstaStats, ok),
      formatStats("CDN   ", cdnStats, ok),
      formatStats("Edge  ", edgeStats, ok),
    ];

    if (failedUrls.length > 0) {
      summaryLines.push("", "── Failed URLs ──");
      failedUrls.forEach((f) =>
        summaryLines.push(`  ✗ ${f.url}\n    Reason: ${f.error}`)
      );
    }

    summaryLines.push(
      "",
      "═══════════════════════════════════════════",
      ""
    );

    const summaryText = summaryLines.join("\n");
    console.log(summaryText);

    // Persist summary to disk for `npm run logs`
    const summaryJson = {
      started: startTime,
      finished: endTime,
      total,
      successful: ok,
      failed: fail,
      kinsta: kinstaStats,
      cdn: cdnStats,
      edge: edgeStats,
      failedUrls,
    };
    fs.writeFileSync(SUMMARY_FILE, JSON.stringify(summaryJson, null, 2));
    log(`Summary saved to ${SUMMARY_FILE}`);

    log("Cache warming cycle completed cleanly.");
  } finally {
    isRunning = false;
  }
}

module.exports = { runWarmer };
